lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp

```diff
@@ -156,18 +168,145 @@ amf_app_desc_t* AmfNasStateManager::get_state(bool read_from_redis) {
+  OAILOG_DEBUG(LOG_AMF_APP, "Clearing state in data store");
+  std::vector<std::string> keys_to_del;
+  keys_to_del.emplace_back(AMF_NAS_STATE_KEY);
+
+  if (redis_client->clear_keys(keys_to_del) != RETURNok) {
+    OAILOG_ERROR(LOG_AMF_APP, "Failed to clear the state in data store");
+    return;
+  }
+}
+
+void put_amf_ue_state(
+    amf_app_desc_t* amf_app_desc_p, imsi64_t imsi64, bool force_ue_write) {
+  if (AmfNasStateManager::getInstance().is_persist_state_enabled()) {
+    if (imsi64 != INVALID_IMSI64) {
+      ue_m5gmm_context_t* ue_context_p = nullptr;
+      amf_ue_ngap_id_t ue_id;
+      get_amf_ue_id_from_imsi(amf_app_desc_p, imsi64, &ue_id);
+      ue_context_p =
+          amf_ue_context_exists_amf_ue_ngap_id((amf_ue_ngap_id_t) ue_id);
+      // Only write MME UE state to redis if force flag is set or UE is in EMM
+      // Registered state
+      if ((ue_context_p && force_ue_write) ||
+          (ue_context_p && ue_context_p->mm_state == REGISTERED_CONNECTED)) {
+        auto imsi_str = AmfNasStateManager::getInstance().get_imsi_str(imsi64);
+        AmfNasStateManager::getInstance().write_ue_state_to_db(
+            ue_context_p, imsi_str);
+      }
+    }
+  }
+}
+
+void AmfNasStateManager::write_ue_state_to_db(
+    const ue_m5gmm_context_t* ue_context, const std::string& imsi_str) {
+#if !MME_UNIT_TEST
+  /* Data store is Redis db. In this case actual call is made to Redis db */
+  StateManager::write_ue_state_to_db(ue_context, imsi_str);
+#else
+  /* Data store is a map defined in AmfClientServicer.In this case call is NOT
+   * made to Redis db */
+  std::string proto_str;
+  magma::lte::oai::UeContext ue_proto = magma::lte::oai::UeContext();
+  AmfNasStateConverter::ue_to_proto(ue_context, &ue_proto);
+  redis_client->serialize(ue_proto, proto_str);
+  std::size_t new_hash = std::hash<std::string>{}(proto_str);
+
+  if (new_hash != this->ue_state_hash[imsi_str]) {
+    std::string key = IMSI_PREFIX + imsi_str + ":" + task_name;
+    if (AMFClientServicer::getInstance().map_imsi_ue_proto_str.insert(
+            key, proto_str) != magma::MAP_OK) {
+      OAILOG_ERROR(
+          log_task, "Failed to write UE state to db for IMSI %s",
+          imsi_str.c_str());
+      return;
+    }
+
+    this->ue_state_version[imsi_str]++;
+    this->ue_state_hash[imsi_str] = new_hash;
+    OAILOG_DEBUG(
+        log_task, "Finished writing UE state for IMSI %s", imsi_str.c_str());
+  }
+#endif
+}
+
+status_code_e AmfNasStateManager::read_ue_state_from_db() {
+#if !MME_UNIT_TEST
+  /* Data store is Redis db. In this case actual call is made to Redis db */
+  // StateManager::read_ue_state_from_db();
+  if (!persist_state_enabled) {
+    return RETURNok;
+  }
+  auto keys = redis_client->get_keys("IMSI*" + task_name + "*");
+  for (const auto& key : keys) {
+    magma::lte::oai::UeContext ue_proto = magma::lte::oai::UeContext();
+    ue_m5gmm_context_t* ue_context_p    = new ue_m5gmm_context_t();
```

Comment:
Move declaration closer to reference, so that no allocation is done if condition in line 270 fails.

Reply:
Comment Addressed